io_scene_niftools/modules/nif_import/geometry/mesh/__init__.py

```diff
@@ -93,42 +93,42 @@ def import_mesh(self, n_block, b_obj):
             if vertex_attributes.vertex_colors:
-                vertex_colors = [vertex.vertex_colors for vertex in vertex_data]
+                vertex_colors = [NifClasses.Color4.from_value(tuple(c / 255.0 for c in vertex.vertex_colors)) for vertex in vertex_data]
             if vertex_attributes.normals:
                 normals = [vertex.normal for vertex in vertex_data]
         elif isinstance(n_block, NifClasses.NiMesh):
-            # if it has a displaylist then we don't know how to process this NiMesh
+            # if it has a displaylist then the vertex data is encoded differently
             displaylist_data = n_block.geomdata_by_name("DISPLAYLIST", False, False)
             if len(displaylist_data) > 0:
                 displaylist = DisplayList(displaylist_data)
-                vertices_info, triangles = displaylist.create_mesh_data(n_block)
+                vertices_info, triangles, weights = displaylist.create_mesh_data(n_block)
                 vertices = vertices_info[0]
                 normals = vertices_info[1]
                 vertex_colors = [NifClasses.Color4.from_value(color) for color in vertices_info[2]]
-                uvs = [[NifClasses.TexCoord.from_value(tex_coord) for tex_coord in vertices_info[3]]]
+                uvs = vertices_info[3]
             else:
                 # get the data from the associated nidatastreams based on the description in the component semantics
                 vertices.extend(n_block.geomdata_by_name("POSITION", sep_datastreams=False))
                 vertices.extend(n_block.geomdata_by_name("POSITION_BP", sep_datastreams=False))
                 triangles = n_block.get_triangles()
                 uvs = n_block.geomdata_by_name("TEXCOORD")
-                if len(uvs) == 0:
-                    uvs = None
-                else:
-                    uvs = [[NifClasses.TexCoord.from_value(tex_coord) for tex_coord in uv_coords] for uv_coords in uvs]
                 vertex_colors = n_block.geomdata_by_name("COLOR", sep_datastreams=False)
                 if len(vertex_colors) == 0:
                     vertex_colors = None
                 else:
                     vertex_colors = [NifClasses.Color4.from_value(color) for color in vertex_colors]
                 normals = n_block.geomdata_by_name("NORMAL", sep_datastreams=False)
                 normals.extend(n_block.geomdata_by_name("NORMAL_BP", sep_datastreams=False))
-                if len(normals) == 0:
-                    normals = None
-                else:
-                    # for some reason, normals can be four-component structs instead of 3, discard the 4th.
-                    if len(normals[0]) > 3:
-                        normals = [(n[0], n[1], n[2]) for n in normals]
+            if len(uvs) == 0:
+                uvs = None
+            else:
+                uvs = [[NifClasses.TexCoord.from_value(tex_coord) for tex_coord in uv_coords] for uv_coords in uvs]
+            if len(normals) == 0:
```

Comment:
could also benefit from np

Reply:
How do you mean?

Reply:
If the extracted normals were stored in an np array, you could just use fancy indexing to drop the 4th coord